pkg/controller/master/virtualmachine/vm_volume_status_controller.go

```diff
@@ -0,0 +1,493 @@
+		}
+		if err != nil {
+			if !errors.IsConflict(err) {
+				logrus.Infof("pvc:(%s/%s) update status to owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vm.Name, fmt.Errorf("%w", err))
+			}
+			return pvc, err
+		}
+	}
+
+	return pvc, nil
+}
+
+// when pvc is used by vm, it is not allowed to delte from webUI
+func (h *VMVolumeStatusController) OnPVCRemove(key string, pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	if pvc == nil {
+		return nil, nil
+	}
+
+	var refOwnerVms []string
+	var err error
+	if refOwnerVms, err = h.getPVCRefOwnerVMs(pvc); err != nil {
+		return nil, err
+	} else if len(refOwnerVms) == 0 {
+		// pvc is not attched to any owner yet
+		return nil, nil
+	}
+
+	// remove pvc status from vm
+	for _, owner := range refOwnerVms {
+		namespace, vmName := ref.Parse(owner)
+		var err error
+		var vm *kubevirtapis.VirtualMachine
+		if vm, err = h.getVM(namespace, vmName); err != nil {
+			if !errors.IsNotFound(err) {
+				logrus.Infof("fail to get owner vm:(%s/%s), give up to remove pvc:%s status, error:%s", namespace, vmName, pvc.Name, fmt.Errorf("%w", err))
+			}
+			continue
+		}
+
+		_, err = h.removeVolumeStatusFromVM(vm, volumeStatusAnnotation{Name: pvc.Name})
+		if err != nil {
+			if !errors.IsConflict(err) {
+				logrus.Infof("pvc:(%s/%s) remove status from owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vmName, fmt.Errorf("%w", err))
+			}
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+// after pvc is detached/unmounted from VM, remove its volume status from VM
+func (h *VMVolumeStatusController) removeVolumeStatusFromDetachedVM(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	var namespace, vmName string
+	var ok bool
+	if namespace, vmName, ok = h.decodeAnnotationDetachedVMString(pvc); !ok {
+		return pvc, nil
+	}
+
+	logrus.Debugf("remove pvc:%s status from previous owner vm:(%s/%s)", pvc.Name, namespace, vmName)
+
+	var err error
+	var vm *kubevirtapis.VirtualMachine
+	if vm, err = h.getVM(namespace, vmName); err != nil {
+		if !errors.IsNotFound(err) {
+			logrus.Infof("fail to get vm:(%s/%s), give up to remove pvc:%s status, error:%s", namespace, vmName, pvc.Name, fmt.Errorf("%w", err))
+		}
+		return pvc, nil
+	}
+
+	if _, err = h.removeVolumeStatusFromVM(vm, volumeStatusAnnotation{Name: pvc.Name}); err != nil {
+		if !errors.IsConflict(err) {
+			logrus.Infof("pvc:(%s/%s) remove status from owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vmName, fmt.Errorf("%w", err))
+		}
+		return pvc, err
+	}
+
+	// delete this temp annotation
+	return h.removePVCAnnotationsDetachedVM(pvc)
+}
+
+func (h *VMVolumeStatusController) getPVC(namespace string, name string) (*corev1.PersistentVolumeClaim, error) {
+	if pvc, err := h.pvcCache.Get(namespace, name); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		} else {
+			return h.pvcClient.Get(namespace, name, metav1.GetOptions{})
+		}
```

Comment:
refer comment in line #97: // pvc may be deleted, but LH volume is still there, in such case, give up updating

when we can not retrieve PVC, wo do not return error, just give up updating status, return nil, otherwise, it may block LH volume from being deleted (#L131) or updated here.

Reply:
You don't need to get from the `pvcClient` again since the cache already contains the latest data with the k8s ListWatch mechanism. Return `nil, nil` if is errors.IsNotFound`, otherwise default to return `pvc, nil`, and in line #97 u should return an error instead of swallowing the other potential errors.

Reply:
You don't need to get from the pvcClient again since the cache already contains the latest data with the k8s ListWatch mechanism
---
A POD normally watches a lot of data, when this POD starts up, those watched data will come in random order. I did see that when LH volume update came, but the related PVC data is not in cache yet. And LH will only update change (into unusable) once. Additional call to h.pvcClient.Get will solve such cases. On the other hand, the "remove" will truly lead to errors.IsNotFound as that object has already been deleted.  There are more "update" than "remove", in "update", the "local cache get" will not miss for most time. It is a trade-off to do so.

This whole fix is a workaround to current implementation, it is more an "observer" other than a "controller", that's why I try not to return error as normal controller, because this may lead to side-effect to current features.  There is no strict data consistency insurance, both upstream and downstream are not knowing this "observer".

L97 is also following this, leave a warning message for trouble-shooting, but still return nil.